scout/server/blueprints/public/controllers.py

```diff
@@ -0,0 +1,182 @@
+import logging
+from functools import reduce
+
+from scout.constants import EVENTS_MAP
+from scout.utils.date import pretty_date
+
+LOG = logging.getLogger(__name__)
+NOF_RECENT_EVENTS = 3
+
+
+"""Controller module for first webview.
+
+Requires: Python 3.7 or greater.
+Reason: Implementation relies on dicts being able to maintain internal order.
+This was previously only supported in OrderedDict().
+"""
+
+
+class CompactEvent:
+    """Class to represent a minimal event to be displayed on Scout's frontpage."""
+
+    def __init__(self, verb, event_type, date):
+        self.verb = verb
+        self.event_type = event_type
+        self.date = date
+        self.count = 1
+
+    def increment(self):
+        self.count = self.count + 1
+        return self
+
+    def __eq__(self, other):
+        if not isinstance(other, CompactEvent):
+            # don't attempt to compare against unrelated types
+            return NotImplemented
+        return (
+            self.verb == other.verb
+            and self.event_type == other.event_type
+            and pretty_date(self.date) == pretty_date(other.date)
+            and self.count == other.count
+        )
+
+    def __repr__(self):
+        return (
+            self.verb + ":" + self.event_type + ":" + str(self.count) + ":" + pretty_date(self.date)
+        )
+
+
+def get_events_of_interest(store, user):
+    """Read event database and compile a list of selected events of interest
+    Args:
+        store: scout.adapter.MongoAdapter
+        user: store.user
+    Returns:
+        events_of_interest: list of dicts{'case', 'link', 'human_readable'}
+        -where human_readable is a list of strings refering to users recent
+         events of interest
+    """
+    events_of_interest = []
+    events_per_case = []
+    cases = recent_cases(user, store)
+    for case in cases:
+        event_list = events_in_case(store, user, case)
+        # only add non-empty event_lists
+        if event_list:
+            events_per_case.append(event_list)
+        else:
+            LOG.debug("no events found for case: {} / {}".format(case, user.email))
+
```

Comment:
I think in the following lines you are making several calls to the store functions to get case's attributes, for instance case display name and owner. Since you already find yourself in a loop over cases ids, why don't you run ONE query to get info over a case dictionary:

case_obj = store.case case_id=case)

Then you'll have all the info you need in the following lines without having to run a database query for each of them